File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.nums = nums
        self.result = list()
        self.construct(list(), [False for _ in range(len(nums))])
        return self.result
    
    def construct(self, permutation, picked):
        if len(permutation) == len(self.nums):
            self.result.append(permutation.copy())
            return
        
        for i in range(len(self.nums)):
            if picked[i]:
                continue
            picked[i] = True
            permutation.append(self.nums[i])
            self.construct(permutation, picked)
            picked[i] = False
            permutation.pop()
        

File: 0046-permutations/0046-permutations.go
var result [][]int
func permute(nums []int) [][]int {
    result = make([][]int, 0)
	calculate(nums, make([]int, 0), make(map[int]bool))
	return result
}

func calculate(nums, currResult []int, picked map[int]bool) {
	if len(currResult) == len(nums) {
        temp := append([]int{}, currResult...)
		result = append(result, temp)
	}

	for _, i := range nums {
		if picked[i] {
			continue
		}
		picked[i] = true
		currResult = append(currResult, i)
		calculate(nums, currResult, picked)
		picked[i] = false
		currResult = currResult[:len(currResult) - 1]
	}
}